### PR TITLE
feat: support load and save hnsw

### PR DESCRIPTION
**Pull request by jemmyshin**
This pr related to this [issue](https://github.com/jina-ai/annlite/issues/132).

Implemented two APIs inside HNSW: `load_index()` and `save_index()`. In this way we can load the HNSW faster than rebuilding the whole graph

### CODE REVIEW

File: annlite/core/index/hnsw/index.py [hunk 1]
```diff
@@ -14,6 +15,7 @@ class HnswIndex(BaseIndex):
     def __init__(
         self,
         dim: int,
+        path_to_load: Optional[str] = None,
```
Comment (JoanFM) [inline annlite/core/index/hnsw/index.py:18]: can u put it at the end so that is less of a breaking change?

**Review by JoanFM — CHANGES_REQUESTED** [review]
Need tests

File: annlite/core/index/hnsw/index.py [hunk 2]
```diff
@@ -34,17 +37,34 @@ def __init__(
...
+                    f'index path: {self.path_to_load} does not exist, a new indexer will be initialized',
+                )
+            self._index.init_index(
+                max_elements=self.capacity,
+                ef_construction=self.ef_construction,
+                M=self.max_connection,
+            )
+            self._index.set_ef(self.ef_search)
+
+    def _load_index(self):
```
Comment (numb3r3) [inline annlite/core/index/hnsw/index.py:63]:
```suggestion
    def load_index(self, index_file: Union[str, Path]):
```

File: annlite/core/index/hnsw/index.py [hunk 3]
```diff
@@ -19,10 +20,12 @@ def __init__(
         ef_construction: int = 200,
         ef_search: int = 50,
         max_connection: int = 16,
+        path_to_load: Optional[str] = None,
```
Comment (numb3r3) [inline annlite/core/index/hnsw/index.py:23]:
```suggestion
        index_file: Optional[Union[str, Path]] = None,
```

File: annlite/core/index/hnsw/index.py [hunk 4]
```diff
@@ -19,10 +20,12 @@ def __init__(
...
         **kwargs,
     ):
         """
         :param dim: The dimensionality of vectors to index
+        :param path_to_load: The path of saved index, will load the indexer from it when provided
```
Comment (numb3r3) [inline annlite/core/index/hnsw/index.py:28]:
```suggestion
        :param index_file: A file-like object or a string containing a file name. 
```

File: annlite/core/index/hnsw/index.py [hunk 5]
```diff
@@ -34,17 +37,34 @@ def __init__(
...
+                max_elements=self.capacity,
+                ef_construction=self.ef_construction,
+                M=self.max_connection,
+            )
+            self._index.set_ef(self.ef_search)
+
+    def _load_index(self):
+        self._index.load_index(self.path_to_load)
+
+    def save_index(self, path):
```
Comment (numb3r3) [inline annlite/core/index/hnsw/index.py:66]:
```suggestion
    def save_index(self, index_file: Union[str, Path]):
```

File: tests/test_hnsw_load_save.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    Xt = np.random.random((n_examples, n_features)).astype(np.float32)
+    return Xt
+
+
+@pytest.fixture
+def build_hnsw(build_data):
+    Xt = build_data
+    hnsw = HnswIndex(dim=512)
+    hnsw.add_with_ids(x=Xt, ids=list(range(len(Xt))))
+    return hnsw
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:23]: yield hnsw

File: tests/test_hnsw_load_save.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_save(tmpdir, build_hnsw):
+    hnsw = build_hnsw
+    hnsw.save_index(os.path.join(tmpdir, 'hnsw.pkl'))
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:28]:
```suggestion
    build_hnsw.save_index(os.path.join(tmpdir, 'hnsw.pkl'))
```

File: tests/test_hnsw_load_save.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+def test_load(tmpdir, build_hnsw):
+    hnsw = build_hnsw
+    hnsw.save_index(os.path.join(tmpdir, 'hnsw.pkl'))
+
+    hnsw_ = HnswIndex(dim=512, path_to_load=os.path.join(tmpdir, 'hnsw.pkl'))
+    assert hnsw_.size == hnsw.size
+
+
+def test_loading_from_wrong_path(tmpdir):
+    hnsw_wrong = HnswIndex(dim=512, path_to_load=os.path.join(tmpdir, 'hnsw_wrong.pkl'))
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:42]: raise an exception if the given index file does not exist.

File: tests/test_hnsw_load_save.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+def test_save(tmpdir, build_hnsw):
+    hnsw = build_hnsw
+    hnsw.save_index(os.path.join(tmpdir, 'hnsw.pkl'))
+
+    assert os.path.exists(os.path.join(tmpdir, 'hnsw.pkl')) == True
+
+
+def test_load(tmpdir, build_hnsw):
+    hnsw = build_hnsw
+    hnsw.save_index(os.path.join(tmpdir, 'hnsw.pkl'))
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:35]: same

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: tests/test_hnsw_load_save.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
+import os
+
+import numpy as np
+import pytest
+
+from annlite.core.index.hnsw import HnswIndex
+
+n_examples = 2000
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:8]: use small num to speedup test

File: tests/test_hnsw_load_save.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+@pytest.fixture
+def build_data():
+    Xt = np.random.random((n_examples, n_features)).astype(np.float32)
+    return Xt
+
+
+@pytest.fixture
+def build_hnsw(build_data):
+    Xt = build_data
+    hnsw = HnswIndex(dim=512)
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:21]:
```suggestion
    hnsw = HnswIndex(dim=n_features)
```

File: tests/test_hnsw_load_save.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_save(tmpdir, build_hnsw):
+    hnsw = build_hnsw
+    hnsw.save_index(os.path.join(tmpdir, 'hnsw.pkl'))
+
+    assert os.path.exists(os.path.join(tmpdir, 'hnsw.pkl')) == True
+
+
+def test_load(tmpdir, build_hnsw):
```
Comment (numb3r3) [inline tests/test_hnsw_load_save.py:33]: can be combined into one test case

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Review by numb3r3 — APPROVED** [review]
LGTM

**Pull request merged**